schism/sample-view.c

```diff
@@ -73,23 +73,33 @@ static void _draw_sample_data_8(struct vgamem_overlay *r,
 		xs = 0;
 		ys = (np - 1) - level;
 		ys = CLAMP(ys, 0, r->height - 1);
-		do {
+
+
+		if (length == 1) {
+			// When we only have one sample to draw, we need to just draw a
+			// horizontal line of whatever that value is, because we don't have
+			// a second sample to draw a line between in the loop below.
+			vgamem_ovl_drawline(r, 0, ys, r->width - 1, ys, SAMPLE_DATA_COLOR);
+		} else {
 			pos += step;
-			co = 0;
-			level=0;
-			do {
-				level+=ceil(data[(pos * inputchans) + cc+co] * nh / (SCHAR_MAX - SCHAR_MIN + 1));
-			} while (co++ < inputchans-outputchans);
-			xe = pos * r->width / length;
-			ye = (np - 1) - level;
-			xe = CLAMP(xe, 0, r->width - 1);
-			ye = CLAMP(ye, 0, r->height - 1);
-			// 'ye' is more or less useless for small samples, but this is much cleaner
-			// code than writing nearly the same loop four different times :P
-			vgamem_ovl_drawline(r, xs, ys, xe, chip ? ys : ye, SAMPLE_DATA_COLOR);
-			xs = xe;
-			ys = ye;
-		} while (pos < length);
+			while (pos < length) {
+				co = 0;
+				level = 0;
+				do {
+					level += ceil(data[(pos * inputchans) + cc + co] * nh / (SCHAR_MAX - SCHAR_MIN + 1));
+				} while (co++ < inputchans - outputchans);
+				xe = pos * r->width / length;
+				ye = (np - 1) - level;
+				xe = CLAMP(xe, 0, r->width - 1);
```

Comment:
What if you just `pos = CLAMP(pos, 0, length); at the start of the `do { } while (pos < length);` loop, then let the existing code handle all cases even length=1, would that not work?

Reply:
so i tried it out, like this:

```c
		pos += step;
		do {
			pos = CLAMP(pos, 0, length - 1);
			co = 0;
			level = 0;
			do {
				level = ceil(data[(pos * inputchans) + cc + co] * nh / (float) (SHRT_MAX - SHRT_MIN + 1));
			} while (co++ < inputchans - outputchans);
			xe = pos * r->width / length;
			ye = (np - 1) - level;
			xe = CLAMP(xe, 0, r->width - 1);
			ye = CLAMP(ye, 0, r->height - 1);
			vgamem_ovl_drawline(r, xs, ys, xe, chip ? ys : ye, SAMPLE_DATA_COLOR);
			xs = xe;
			ys = ye;
			pos += step;
		} while (pos < length);
```

This renders the big samples ok but it doesn't render anything for length=1. The reason is because the X coordinate on screen `xe` is based on `pos`. Since `pos` is 0 (there is only one sample so it is clamped to 0), then `xe` is 0 and it doesn't render anything because `xs == xe`.

I did `pos = CLAMP(pos, 0, length - 1)` instead of `pos = CLAMP(pos, 0, length)` because otherwise you would have the out of bounds memory access still.

Reply:
Hmm, this function just stinks as-is.  I'm a bit short on time to dig deeper into this right now, maybe @mrpapersonic can take a look.

But thanks for shining light on this bug.  I'm leaning towards we take the opportunity to clean it up properly if we can.

Reply:
yeah i was thinking the same. i was hoping to get in a fix to avoid the crashing that can happen at the moment but it definitely could stand for a proper cleanup also.